src/box/lua/net_cluster.lua

```diff
@@ -0,0 +1,466 @@
+            pcall(watcher._func, key, value, replicaset_name)
+            state.is_executing = nil
+        end
+        if state.wait_cond then
+            state.wait_cond:signal()
+        end
+    end
+end
+
+-- Subscribe watcher to one particular replicaset.
+local function watch_replicaset(watcher, replicaset)
+    if not replicaset_comply_roles(replicaset, watcher._opts.roles) then
+        return nil
+    end
+    local replicaset_name = replicaset.name
+    if not watcher._states[replicaset_name] then
+        watcher._states[replicaset_name] = {}
+    end
+    local wwfunc = wrap_watcher_function(watcher, replicaset_name)
+    local conn = acquire_replicaset_connection(watcher._cluster, replicaset)
+    local handle = conn:watch_leader(watcher._key, wwfunc)
+    watcher._states[replicaset_name].handle = handle
+end
+
+-- Load cluster config. Internal method, expects config to be correct.
+local function reload_config_impl(cluster, config)
+    -- Disconnect.
+    for watcher in pairs(cluster._watchers) do
+        for _, state in pairs(watcher._states) do
+            if state.handle then
+                state.handle:unregister()
+                state.handle = nil
+            end
+        end
+    end
+    for _, replicaset in pairs(cluster._replicasets) do
+        close_replicaset_connection(replicaset)
+    end
+    table.clear(cluster._replicasets)
+
+    -- Build replicaset map.
+    for replicaset_name, replicaset_cfg in pairs(config) do
+        local connect_cfg = replicaset_cfg.connect_cfg
+        connect_cfg = connect_cfg and table.deepcopy(connect_cfg)
+        local roles = {}
+        if replicaset_cfg.roles then
+            for k, v in pairs(replicaset_cfg.roles) do
+                if type(k) == 'number' and type(v) == 'string' then
+                    roles[v] = v
+                elseif type(k) == 'string' and v then
+                    roles[k] = k
+                end
+            end
+        end
+        local replicaset = {
+            name = replicaset_name,
+            connect_cfg = connect_cfg,
+            roles = roles,
+        }
+        cluster._replicasets[replicaset_name] = replicaset
+    end
+
+    -- Connect and subscribe if necessary.
+    for _, replicaset in pairs(cluster._replicasets) do
+        for watcher in pairs(cluster._watchers) do
+            watch_replicaset(watcher, replicaset)
+        end
+    end
+end
+
+local function collect_config()
+    local config = require('config')
+    local replicasets = {}
+
+    -- Find own group.
+    local own_group_name
+    for _, info in pairs(config:instances()) do
+        if info.instance_name == box.info.name then
+            own_group_name = info.group_name
+            break
+        end
+    end
+
+    -- Find replicasets of own group.
+    for _, info in pairs(config:instances()) do
+        if info.group_name ~= own_group_name then
+            goto continue
+        end
```

Comment:
I see no reason, why we should work only within our own group. Also, as I see on a brief glance, this constraint is not mentioned in comments (which hopefully will be transformed later into the documentation).